pallets/msa/src/lib.rs

```diff
@@ -1128,18 +1130,30 @@ impl<T: Config> Pallet<T> {
 	/// * [`Error::DelegationNotFound`]
 	/// * [`Error::DelegationRevoked`]
 	/// * [`Error::SchemaNotGranted`]
+	/// * [`Error::CannotPredictValidityPastCurrentBlock`]
 	///
 	pub fn ensure_valid_schema_grant(
 		provider: Provider,
 		delegator: Delegator,
 		schema_id: SchemaId,
+		block_number: T::BlockNumber,
 	) -> DispatchResult {
-		let provider_info = Self::ensure_valid_delegation(provider, delegator, None)?;
+		let provider_info = Self::ensure_valid_delegation(provider, delegator, Some(block_number))?;
+
+		let schema_permission_revoked_at_block_number = provider_info
+			.schema_permissions
+			.get(&schema_id)
+			.ok_or(Error::<T>::SchemaNotGranted)?;
+
+		if *schema_permission_revoked_at_block_number == T::BlockNumber::zero() {
+			return Ok(())
+		}
 
 		ensure!(
-			provider_info.schema_permissions.contains_key(&schema_id),
+			block_number <= *schema_permission_revoked_at_block_number,
 			Error::<T>::SchemaNotGranted
 		);
```

Comment:
Suggestion: This feels like a perfect place for a match statement to make it clear the three outcomes.

```
match schema_permission_revoked_at_block_number {
	x if *x == T::BlockNumber::zero() => Ok(()),
	x if *x < block_number => Ok(()),
	_ => Err(Error::<T>::SchemaNotGranted.into())
} 
```

Bit sad we cannot just use ranges (`1..=block_number => Ok`), but BlockNumber would have to be converted first and that is more of a pain.

Reply:
Interesting.